src/plonky2_verifier/chip/native_chip/arithmetic_chip.rs

```diff
@@ -0,0 +1,610 @@
+            let right_x = p.clone() * q_x.clone() + r_x.clone();
+            let right_y = p.clone() * q_y.clone() + r_y.clone();
+            vec![
+                s_ext.clone() * (left_x - right_x),
+                s_ext.clone() * (left_y - right_y),
+            ]
+        });
+
+        q_limbs.iter().for_each(|limb| {
+            meta.lookup("q_limbs range check", |meta| {
+                let l = meta.query_advice(*limb, Rotation::cur());
+                vec![(l, table)]
+            });
+        });
+        r_limbs.iter().for_each(|limb| {
+            meta.lookup("r_limbs range check", |meta| {
+                let l = meta.query_advice(*limb, Rotation::cur());
+                vec![(l, table)]
+            });
+        });
+        ArithmeticChipConfig {
+            a,
+            b,
+            c,
+            q,
+            r,
+            q_limbs,
+            r_limbs,
+            table,
+            instance,
+            constant,
+            s_limb,
+            s_range,
+            s_base,
+            s_ext,
+            _marker: PhantomData,
+        }
+    }
+}
+
+pub struct AssignedArithmetic<F: PrimeField> {
+    pub a: AssignedCell<F, F>,
+    pub b: AssignedCell<F, F>,
+    pub c: AssignedCell<F, F>,
+    pub r: AssignedCell<F, F>,
+}
+
+pub struct AssignedArithmeticExt<F: PrimeField> {
+    pub a: [AssignedCell<F, F>; 2],
+    pub b: [AssignedCell<F, F>; 2],
+    pub c: [AssignedCell<F, F>; 2],
+    pub r: [AssignedCell<F, F>; 2],
+}
+
+#[derive(Clone)]
+pub enum Term<'a, F: PrimeField> {
+    Assigned(&'a AssignedCell<F, F>),
+    Unassigned(Value<F>),
+}
+
+#[derive(Clone)]
+pub enum TermExt<'a, F: PrimeField> {
+    Assigned(&'a [AssignedCell<F, F>; 2]),
+    Unassigned([Value<F>; 2]),
+}
+
+#[derive(Clone, Debug)]
+pub struct ArithmeticChip<F: PrimeField> {
+    config: ArithmeticChipConfig<F>,
+}
+
+impl<F: PrimeField> ArithmeticChip<F> {
+    pub fn new(config: &ArithmeticChipConfig<F>) -> Self {
+        ArithmeticChip {
+            config: config.clone(),
+        }
+    }
+
+    pub fn expose_public(
+        &self,
+        mut layouter: impl Layouter<F>,
+        value: AssignedCell<F, F>,
+        row: usize,
+    ) -> Result<(), Error> {
+        layouter.constrain_instance(value.cell(), self.config.instance, row)
+    }
+
+    pub fn assert_equal(
+        &self,
+        ctx: &mut RegionCtx<'_, F>,
+        a: &AssignedCell<F, F>,
+        b: &AssignedCell<F, F>,
+    ) -> Result<(), Error> {
+        ctx.constrain_equal(a.cell(), b.cell())?;
+        Ok(())
+    }
+
+    pub fn assert_equal_ext(
+        &self,
+        ctx: &mut RegionCtx<'_, F>,
+        a: &[AssignedCell<F, F>; 2],
+        b: &[AssignedCell<F, F>; 2],
+    ) -> Result<(), Error> {
+        for i in 0..2 {
+            ctx.constrain_equal(a[i].cell(), b[i].cell())?;
+        }
+        Ok(())
+    }
+
+    //If the given constant is already assigned, return the cell. If not, assign it.
+    pub fn assign_constant(
```

Comment:
In my understanding, constants do not need to be range-checked at the circuit level. Since constants cannot be changed after the circuit is built, it is sufficient to directly verify that the constants are within range at build time.